src/blocks/libvirt.rs

```diff
@@ -0,0 +1,149 @@
+use virt::connect::Connect;
+
+use std::time::Duration;
+
+use crossbeam_channel::Sender;
+use serde_derive::Deserialize;
+
+use crate::blocks::{Block, ConfigBlock, Update};
+use crate::config::SharedConfig;
+use crate::de::deserialize_duration;
+use crate::errors::*;
+use crate::formatting::value::Value;
+use crate::formatting::FormatTemplate;
+use crate::scheduler::Task;
+use crate::widgets::text::TextWidget;
+use crate::widgets::I3BarWidget;
+use crate::widgets::State;
+
+pub struct Libvirt {
+    id: usize,
+    text: TextWidget,
+    format: FormatTemplate,
+    update_interval: Duration,
+    qemu_conn: Connect,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct LibvirtConfig {
+    /// Update interval in seconds
+    #[serde(deserialize_with = "deserialize_duration")]
+    pub interval: Duration,
+
+    /// Format override
+    pub format: FormatTemplate,
+
+    /// URL to QEMU
+    pub qemu_url: String,
+}
+
+impl Default for LibvirtConfig {
+    fn default() -> Self {
+        Self {
+            interval: Duration::from_secs(5),
+            format: FormatTemplate::default(),
+            qemu_url: "qemu:///system".to_string(),
+        }
+    }
+}
+
+impl ConfigBlock for Libvirt {
+    type Config = LibvirtConfig;
+
+    fn new(
+        id: usize,
+        block_config: Self::Config,
+        shared_config: SharedConfig,
+        _: Sender<Task>,
+    ) -> Result<Self> {
+        let text = TextWidget::new(id, 0, shared_config)
+            .with_text("vms")
+            .with_icon("virtual-machine")?;
+
+        Ok(Libvirt {
+            id,
+            text,
+            format: block_config.format.with_default("{running}")?,
+            update_interval: block_config.interval,
+            qemu_conn: Connect::open_read_only(&block_config.qemu_url)
+                .block_error("virt", "error when connecting to libvirt")?,
+        })
+    }
+}
+
+impl Block for Libvirt {
+    fn update(&mut self) -> Result<Option<Update>> {
+        // Check if the connection is still active and re-instatiate it if it's not
+        if !self
+            .qemu_conn
+            .is_alive()
+            .block_error("virt", "error when getting connection status to libvirt")?
+        {
+            self.qemu_conn = Connect::open_read_only(
+                &self
+                    .qemu_conn
+                    .get_uri()
```

Comment:
This method returns the same string as `Config::qemu_url`, right? Let's just store it ourselves in `Libvirt`?

Reply:
Sure, why not, I figured I was being clever, but a very tiny bit of memory overhead vs a CPU overhead is probably worthwhile.

Reply:
Looking at the implementation of `get_uri()` ([here](https://docs.rs/virt/latest/src/virt/connect.rs.html#641-649) and [here](https://docs.rs/virt/latest/src/virt/lib.rs.html#91-93)), It seems that every call to `get_uri` first allocates a string (on the C side), then allocates a second string (on the Rust side), then deallocates the first string and only then returns an owner Rust string (which shortly after will get deallocated). So, there is a point of time when both strings exist, so we can avoid calling `get_uri` without feeling bad for wasting resources xD